src/What4/ExprHelpers.hs

```diff
@@ -181,23 +190,139 @@ anyPred ::
+  VarBinds sym ->
+  [Pair (W4.BoundVar sym) (W4.SymExpr sym)]
+flattenVarBinds (VarBinds binds) = concat $ map go (MapF.toList binds)
+  where
+    go :: Pair (W4.SymExpr sym) (SetF (W4.BoundVar sym)) -> [Pair (W4.BoundVar sym) (W4.SymExpr sym)]
+    go (Pair e bvs) = map (\bv -> Pair bv e) $ SetF.toList bvs
+
+toBindings ::
+  forall sym t solver fs.
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  VarBinds sym ->
+  Pair (Ctx.Assignment (W4.BoundVar sym)) (Ctx.Assignment (W4.SymExpr sym))
+toBindings varBinds = toAssignmentPair (flattenVarBinds varBinds)
+
+-- | Traverse an expression and rewrite any sub-expressions according to the given
+-- replacement function, rebuilding as necessary
+rewriteSubExprs ::
+  forall sym t solver fs tp.
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  (forall tp'. W4B.Expr t tp' -> Maybe (W4B.Expr t tp')) ->
+  W4B.Expr t tp ->
+  IO (W4B.Expr t tp)
+rewriteSubExprs sym f e = do
+  cache <- freshVarBindCache
+  rewriteSubExprs' sym cache f e
+
+data VarBindCache sym where
+  VarBindCache :: sym ~ W4B.ExprBuilder t solver fs => W4B.IdxCache t (Tagged (VarBinds sym) (W4B.Expr t)) -> VarBindCache sym
+
+freshVarBindCache ::
+  forall sym t solver fs.
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  IO (VarBindCache sym)
+freshVarBindCache = VarBindCache <$> W4B.newIdxCache
+
+rewriteSubExprs' ::
+  forall sym tp.
+  sym ->
+  VarBindCache sym ->
+  (forall tp'. W4.SymExpr sym tp' -> Maybe (W4.SymExpr sym tp')) ->
+  W4.SymExpr sym tp ->
+  IO (W4.SymExpr sym tp)
+rewriteSubExprs' sym (VarBindCache cache) = rewriteSubExprs'' sym cache
+
+rewriteSubExprs'' ::
+  forall sym t solver fs tp.
+  sym ~ (W4B.ExprBuilder t solver fs) =>
+  sym ->
+  W4B.IdxCache t (Tagged (VarBinds sym) (W4B.Expr t))  ->
+  (forall tp'. W4B.Expr t tp' -> Maybe (W4B.Expr t tp')) ->
+  W4B.Expr t tp ->
+  IO (W4B.Expr t tp)
+rewriteSubExprs'' sym cache f e_outer = do
+  let
+    go :: forall tp'. W4B.Expr t tp' -> CMW.WriterT (VarBinds sym) IO (W4B.Expr t tp')
+    go e = idxCacheEvalWriter cache e $ case f e of
+      Just e' -> do
+        bv <- IO.liftIO $ W4.freshBoundVar sym W4.emptySymbol (W4.exprType e')
+        CMW.tell $ VarBinds $ MapF.singleton e' (SetF.singleton bv)
+        return $ W4.varExpr sym bv
+      Nothing -> case e of
+        W4B.AppExpr a0 -> do
+          a0' <- W4B.traverseApp go (W4B.appExprApp a0)
+          if (W4B.appExprApp a0) == a0' then return e
+          else IO.liftIO $ W4B.sbMakeExpr sym a0'
+        W4B.NonceAppExpr a0 -> do
+          a0' <- TFC.traverseFC go (W4B.nonceExprApp a0)
+          if (W4B.nonceExprApp a0) == a0' then return e
+          else IO.liftIO $ W4B.sbNonceExpr sym a0'
+        _ -> return e
+  (e', binds) <- CMW.runWriterT (go e_outer)
+  Pair vars vals <- return $ toBindings binds
+  fn <- W4.definedFn sym W4.emptySymbol vars e' W4.AlwaysUnfold
+  W4.applySymFn sym fn vals >>= fixMux sym
+
+
+type ExprBindings sym = MapF.MapF (W4.SymExpr sym) (W4.SymExpr sym)
```

Comment:
Can we just say a bit here about what is in the map? Are there any interesting invariants here?